dlrover/python/elastic_agent/torch/ckpt_saver.py

```diff
@@ -722,91 +797,139 @@ def _commit_checkpoint(
+        else:
+            self._to_save_queue = None
+        # lock for shared memory
+        local_shard_num = self.get_local_shard_num()
+        self.local_shard_id = self._local_rank % local_shard_num
+        lock_name = _SHM_LOCK_NAME_PREFIX + str(self.local_shard_id)
+        self._shm_lock = SharedLock(name=lock_name, create=False)
+        self._shm_handler = SharedMemoryHandler(
+            self.local_shard_id, host=False
+        )
+        self._notify_agent_to_create_saver()
 
-    def save_shm_to_storage(self):
-        """
-        Save the state dict in the shared memory into the storage. The agent
-        can call the method to save the state dict into the storage if the
-        training process fails or the agent wants to restart training
-        processes.
-        """
-        logger.info("Agent terminated, save latest checkpoint to storage")
-        for handler in self._shm_handlers:
-            if handler.empty():
-                return
+    def __del__(self):
+        self.close()
 
-        if self._writing_storage:
-            logger.info("Saver is writing to storage, waiting...")
-            start = time.time()
-            while self._writing_storage:
-                time.sleep(10)
-                elapsed_time = time.time() - start
-                if elapsed_time > 120:
-                    logger.error("Saver writing to storage, timeout")
-                    return
+    def close(self):
+        self._shm_handler.close()
 
-        if not self._acquire_all_locks():
-            # The training process does not release the lock because it fails
-            # when writing the state dict into the shared memory. The shared
-            # memory may be dirty and the saver cannot save it to the storage.
+    def _notify_agent_to_create_saver(self):
+        if self._local_rank != 0:
             return
+        if self._restart_count > 0:
+            # Only local rank 0 notify to initialize the saver in
+            # the main process at the first start.
+            # Avoid the lock is locked by a failed process.
+            self._shm_lock.release()
+            return
+        queue = SharedQueue(name="factory")
+
+        local_shard_num = self.get_local_shard_num()
+        global_shard_num = self.get_global_shard_num()
+        clazz = self.get_saver_class()
+        class_meta = SaverClassMeta(
+            module_path=clazz.__module__,
+            class_name=clazz.__name__,
+            init_args={
+                "checkpoint_dir": self.checkpoint_dir,
+                "local_shard_num": local_shard_num,
+                "global_shard_num": global_shard_num,
+            },
+        )
 
-        step = self._shm_handlers[0].get_iteration_step()
-        # we need to release the locks before save to storage
+        queue.put(class_meta)
+        queue.unlink()
 
-        for lock in self._shm_lock:
-            lock.release()
+    @timer
+    def save_to_memory(self, step, state_dict, path=""):
```

Comment:
we may need a block parameter  here, when user is calling save_to_storage, we should update memory synchronously, or we may loss the step user wanted.

Reply:
save_to_storage will firstly call save_to_memory and then notify an event to the agent to save.

Reply:
yes，but if the agent is writing to storage now, the save_to_memory can't acquire the lock, then the memory sync will be skipped

Reply:
It is reasonable to skip `save_to_memory` if the agent is writing to the storage.  Otherwise, the `save_to_memory` may block the training too long if the time to write to the storage is long.